File: algorithms/[1600]皇位继承顺序/solution.ts
// topics = ["设计", "树"]

class ThroneInheritance {
  edges: Map<string, string[]>;
  dead: Set<string>;
  king: string;
  constructor(kingName: string) {
    this.edges = new Map();
    this.dead = new Set();
    this.king = kingName;
  }

  birth(parentName: string, childName: string): void {
    if (!this.edges.has(parentName)) {
      this.edges.set(parentName, []);
    }
    this.edges.get(parentName)!.push(childName);
  }

  death(name: string): void {
    this.dead.add(name);
  }

  /**
   * 多叉树前序遍历
   * time O(n), space O(n), n 为总人数
   */
  getInheritanceOrder(): string[] {
    const res: string[] = [];

    const preorder = (name: string): void => {
      if (!this.dead.has(name)) {
        res.push(name);
      }
      if (this.edges.has(name)) {
        for (const childName of this.edges.get(name)!) {
          preorder(childName);
        }
      }
    };

    preorder(this.king);

    return res;
  }
}

/**
 * Your ThroneInheritance object will be instantiated and called as such:
 * var obj = new ThroneInheritance(kingName)
 * obj.birth(parentName,childName)
 * obj.death(name)
 * var param_3 = obj.getInheritanceOrder()
 */

File: algorithms/[1600]皇位继承顺序/solution.py
# topics = ["设计", "树"]

from collections import defaultdict
from typing import DefaultDict, List, Set


class ThroneInheritance:
    def __init__(self, kingName: str):
        self.edges: DefaultDict[str, List[str]] = defaultdict(list)
        self.dead: Set[str] = set()
        self.king = kingName

    def birth(self, parentName: str, childName: str) -> None:
        self.edges[parentName].append(childName)

    def death(self, name: str) -> None:
        self.dead.add(name)

    def getInheritanceOrder(self) -> List[str]:
        """
        多叉树前序遍历
        time O(n), space O(n), n 为总人数
        """
        res: List[str] = []

        def preorder(name: str) -> None:
            if name not in self.dead:
                res.append(name)
            if name in self.edges:
                for childName in self.edges[name]:
                    preorder(childName)

        preorder(self.king)

        return res


# Your ThroneInheritance object will be instantiated and called as such:
# obj = ThroneInheritance(kingName)
# obj.birth(parentName,childName)
# obj.death(name)
# param_3 = obj.getInheritanceOrder()
